[rocThrust] Adding docs to unique_ptr.h

marsavicc opened the pull request:
## Motivation

Add documentation for `thrust::unique_ptr` to improve API usability and maintainability. The documentation follows the structure and style of cppreference.com (https://en.cppreference.com/w/cpp/memory/unique_ptr.html) while incorporating rocThrust-specific details about device memory management.

## Test Plan

Documentation changes only - no functional changes to test.

yungshengtu commented:
@marsavicc I noticed that after adding the new documentation, some of the previously undocumented warnings are still present, and a few additional warnings have appeared. The docs build seems to fail with errors at the end. Just a heads-up in case it hasn’t been spotted yet.

FYI, at the moment, in rocThrust, it looks like only unique_ptr.h is producing warnings and errors when building the docs.

marsavicc commented:
> @marsavicc I noticed that after adding the new documentation, some of the previously undocumented warnings are still present, and a few additional warnings have appeared. The docs build seems to fail with errors at the end. Just a heads-up in case it hasn’t been spotted yet.
> 
> FYI, at the moment, in rocThrust, it looks like only unique_ptr.h is producing warnings and errors when building the docs.

@yungshengtu Sorry for errors and warnings, I didn't follow the guide on how to build docs properly and didn't notice these problems. I fixed the warnings for missing docs. 

Regarding the error that was popping up, it seems to me that Doxygen 1.9.4's C++ parser doesn't recognize modern C++ attributes (`__attribute__((trivial_abi))` and `[[no_unique_address]]`). When it encounters these, it fails to parse the class structure correctly, causing malformed XML that crashes Sphinx/Breathe. I added these attributes to Doxyfile `PREDEFINED` section and it isn't causing problems anymore. 

Also because we added a new macro for this implementation, Sphinx's C++ domain parser didn't understand our macro `THRUST_CONSTEXPR_SINCE_CXX23` used in function signatures and it generated warnings about invalid C++ declarations. I added our macro to `cpp_id_attributes` in `conf.py` to treat it as an identifier attribute (like `inline` or `constexpr`) and ignore it during parsing. 

Is this fix valid?

yungshengtu commented:
@marsavicc Yes, it works! Thanks for the update!

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -16,22 +22,61 @@
+  /*! \brief Deletes the object pointed to by \p ptr.
+   *
+   *  For non-trivially destructible types, invokes the destructor on the device
+   *  before deallocating memory.
```
Should we simplify this description by removing the caveat about trivially destructible types? It's _mostly_ just an optimization that the user doesn't need to know/think about, but then again, it might have implications for some edge case uses. E.g. if a user tries to do something fancy with CUDA streams and events to wait on the destruction or something like that.

marsavicc replied:
I simplified it. 

I kept mentions of trivially destructible only where we explain why we have 2 `default_delete` implementations and for constructors in the array form.

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -103,6 +175,10 @@ struct default_delete<T[], std::enable_if_t<!std::is_trivially_destructible_v<T>
+   *
+   *  \return The array size.
+   */
   THRUST_HOST size_t size() const
```
Why is this a public method? Is it used anywhere other than the copy/converting constructor? This seems like we're exposing implementation details for no reason, introducing the possibility of users depending on it and locking us into a specific implementation for no particular reason. Could we just remove it entirely and have the copy/converting constructor access `m_size` directly?

marsavicc replied:
Yep, I removed the method.

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -112,25 +188,49 @@ struct default_delete<T[], std::enable_if_t<!std::is_trivially_destructible_v<T>
+ *  does NOT store the element count. Their destruction is a no-op, so only the raw
+ *  deallocation is required. Omitting the size keeps this deleter specialization an
+ *  empty (zero-size) type, allowing \p unique_ptr<T[]> instantiations that use it to
+ *  remain a zero-cost abstraction (the deleter need not increase the overall object size).
```
*Technically* until the dust settles around `trivially_relocatable`, `unique_ptr<T[]>` is not _always_ a zero-cost abstraction, even for trivially destructible types as it _may_ prevent a bitwise relocation of the pointer itself. Also wouldn't hurt to elaborate a bit more that we're talking about increasing the size of the `unique_ptr<T[]>` compared to a raw pointer.
```suggestion
 *  remain a low/zero-cost abstraction (the default deleter need not increase the overall size of unique_ptr<T[]> above the size of a simple pointer `T *`).
```

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -60,21 +105,48 @@ struct default_delete<T, std::enable_if_t<!std::is_array_v<T>>>
   THRUST_HOST constexpr default_delete(size_t n = 0) noexcept
       : m_size(n){};
 
+  /*! \brief Converting constructor from compatible array deleter.
```
```suggestion
  /*! \brief Copy and converting constructor from compatible array deleter.
```

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -112,25 +188,49 @@ struct default_delete<T[], std::enable_if_t<!std::is_trivially_destructible_v<T>
+   */
   THRUST_HOST constexpr default_delete(size_t = 0) noexcept {};
 
+  /*! \brief Converting constructor from compatible deleter.
```
```suggestion
  /*! \brief Copy and converting constructor from compatible deleter.
```

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -182,15 +282,64 @@ struct unique_ptr_deleter_sfinae<Deleter&>
+ *
+ *  The object is disposed of by calling `get_deleter()(get())`. The default deleter,
+ *  `thrust::default_delete`, deallocates the memory using `thrust::device_free`.
+ *  For non-trivially destructible types, Deleter invokes the destructor of the
```
Similar to my above comment, we might want to remove most of the references to trivially-destructible. Invoking the destructor for a trivially-destructible type is a no-op, so it shouldn't make a difference to the user whether we call it or we don't.

Obviously we still need to keep the mentions of trivially-destructible vs not anywhere it has API implications (e.g. to explain why we have 2 versions of `default_delete`, or why some constructors for unique_ptr<T[]> are only available for trivially-destructible types), but most of the rest can probably be removed.

marsavicc replied:
Answered above.

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -60,21 +105,48 @@ struct default_delete<T, std::enable_if_t<!std::is_array_v<T>>>
+/*! \brief Default deleter specialization for arrays of non-trivially destructible types.
+ *
+ *  This specialization handles deletion of arrays where element destructors must be called.
+ *  Stores the array size to ensure all element destructors are properly invoked on the device.
```
"Stores the array size to ensure all element destructors are properly invoked on the device." - this is mostly an implementation detail that doesn't need to be included in the documentation for users.

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -414,12 +653,54 @@ class __attribute__((trivial_abi)) unique_ptr
+ *  provides `operator[]` for element access.
+ *
+ *  \par Array Destruction Behavior
+ *  The array form has different constructor requirements and performance characteristics
+ *  based on the element type's destructibility:
+ *
+ *  **Trivially-destructible types** (e.g., `int`, `float`, POD types):
+ *  - Constructors do NOT require array size parameter
+ *  - Deleter is zero-size (no memory overhead)
+ *  - Destruction only deallocates memory (no destructors to call)
+ *  - Example: `unique_ptr<int[]> ptr(device_malloc<int>(100));`
+ *
+ *  **Non-trivially-destructible types** (e.g., types with custom destructors):
+ *  - Constructors REQUIRE array size parameter
+ *  - Deleter stores the size (small memory overhead: one `size_t`)
+ *  - Destruction calls destructor for each element on device before deallocation
+ *  - Example: `unique_ptr<MyClass[]> ptr(device_new<MyClass>(100), 100);`
```
This is maybe another exception to my above comments suggesting we remove references to trivially_destructible.

marsavicc replied:
Answered above.

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -632,33 +994,57 @@ class __attribute__((trivial_abi)) unique_ptr<T[], D>
+   *
+   *  For device arrays, accessing an element from host code triggers a copy
+   *  from device to host memory.
+   *
```
Can we add a warning about performance here? Maybe something like "Using this when iterating over the array is STRONGLY discouraged as the excessive number of single-element copies usually results in poor performance"

DanielMcIntosh commented on projects/rocthrust/thrust/unique_ptr.h:
```diff
@@ -848,13 +1355,38 @@ template <class T, class D>
   return unique_ptr<T>(thrust::device_new<T>(p, T(std::forward<Args>(args)...), 1));
 }
 
+/*! \brief Constructs an array of objects of type \p T in device memory and wraps it in a \p unique_ptr.
```
We probably want to mention what kind of initialization is used to initialize the elements - value-initialization, default-initialization, etc.
